### PR TITLE
SNOW-1370306: local testing fix float cast precision loss and pandas timestamp related issues

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-NNNNNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

- Fixed a bug that integer precision of large value gets lost when converted to pandas DataFrame.
- Fixed a bug that the schema of datetime object is wrong when create DataFrame from a pandas DataFrame.
- raised error that pandas timestamp containing tz info is not supported
- consolidated pandas related tests

### CODE REVIEW

File: src/snowflake/snowpark/mock/_pandas_util.py [hunk 1]
```diff
@@ -202,17 +207,23 @@ def convert_to_python_obj(obj):
...
+                )
+            else:
+                data_type = TimestampType(TimestampTimeZone.NTZ)
         elif pandas_type.type == numpy.float64:
             data_type = DoubleType()
         elif isinstance(pandas_type, (pd.Float32Dtype, pd.Float64Dtype)):
             data_type = DoubleType()
-        elif (
-            pandas_type.type == numpy.int64
-            or pandas_type.type == numpy.datetime64
```
Comment (sfc-gh-jrose) [inline src/snowflake/snowpark/mock/_pandas_util.py:213]: Is datetime64 covered by a different branch now?

File: src/snowflake/snowpark/mock/_functions.py [hunk 2]
```diff
@@ -488,8 +488,15 @@ def mock_to_decimal(
         [x] If the variant contains JSON null value, the output is NULL.
     """
 
+    def check_int(s):
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/_functions.py:491]:
```suggestion
    def is_str_int(s):
```

File: src/snowflake/snowpark/mock/_functions.py [hunk 3]
```diff
@@ -816,7 +823,7 @@ def mock_timestamp_ntz(
     result = _to_timestamp(column, fmt, try_cast, enforce_ltz=True)
     # Cast to NTZ by removing tz data if present
     return ColumnEmulator(
-        data=[x.replace(tzinfo=None) for x in result],
+        data=[x.replace(tzinfo=None) if x is not None else x for x in result],
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/_functions.py:826]: Can this be rewritten using `try_convert`?

File: tests/integ/test_dataframe.py [hunk 4]
```diff
@@ -1578,7 +1578,7 @@ def test_create_dataframe_with_semi_structured_data_types(session):
 
 @pytest.mark.skipif(
     "config.getoption('local_testing_mode', default=False)",
-    reason="SNOW-1368516 create dataframe from pandas dataframe with datetime columns has wrong schema",
+    reason="SNOW-1439717 create dataframe from pandas dataframe containing timestamp with tzinfo is not supported.",
```
Comment (sfc-gh-stan) [inline tests/integ/test_dataframe.py:1581]: Please document this as a known limitation.

File: tests/integ/test_pandas_to_df.py [hunk 5]
```diff
@@ -549,3 +564,349 @@ def test_write_to_different_schema(session, local_testing_mode):
     finally:
         if not local_testing_mode:
             Utils.drop_schema(session, test_schema_name)
+
+
+@pytest.mark.localtest
```
Comment (sfc-gh-stan) [inline tests/integ/test_pandas_to_df.py:569]: This marker is no longer needed.

Comment (sfc-gh-stan) [reply tests/integ/test_pandas_to_df.py:569]: I have a PR that removes these: https://github.com/snowflakedb/snowpark-python/pull/1562

**Review by sfc-gh-stan — CHANGES_REQUESTED** [review]

File: src/snowflake/snowpark/mock/_functions.py (hunk 3 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/_functions.py]: yes, it's already updated in another PR

File: src/snowflake/snowpark/mock/_pandas_util.py (hunk 1 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/_pandas_util.py:213]: this, this is moved to line 210 where DateTime is handled

File: tests/integ/test_dataframe.py (hunk 4 above)
Comment (sfc-gh-aling) [reply tests/integ/test_dataframe.py:1579]: good callout, I have added that into the goole doc

**Review by sfc-gh-jrose — APPROVED** [review]

**Review by sfc-gh-stan — APPROVED** [review]

**Pull request merged**